### PR TITLE
Storage trait changes

### PR DESCRIPTION
**Pull request by sandreae**
Part of one of the "code love" NLNet milestones, deals with some left over issues around improving the API of our storage traits. Bulk of the changes were being tracked here: [#149](https://github.com/p2panda/aquadoggo/issues/149)

Looking at these now after some time I spotted one or two implementation decisions which could be improved on, in particular in the `DocumentStore` API which inadvertently introduced a concept of a cache layer, which is not wanted/needed in p2panda-rs. More on that here: #468 .

The general name of the game is: make it simple, less is more, only what we really need.

I want it to be as easy as possible to implement these storage traits for other backends while still satisfying our storage requirements for our core data types `Entry` and `Operation`. 

## Changes
- [x] remove `StorageLog` struct/trait in favour of passing individual parameters
- [x] general renaming of badly named methods
- [x] remove `next_log_id` method which was too smart and has been moved to the `validation` module
- [x] move remaining top level method to it's store
- [x] remove `StorageProvider` trait completely (!) in favour of requiring individual storage traits which encourages separation of concerns.
- [x] remove any concept of "caching" documents and views from `DocumentStore` only leave basic "get" methods which have default trait methods which require `OperationStore` and internally materialise docs each time they're called (inefficient, but no cache!)
- [x] already introduce `WithId<T>` trait
- [x] refactor out test database structs in favour of manually populating `MemoryStore` in tests
- [x] remove use of blanket conversion trait implementations for `Entry` and `EncodedEntry` (this is, apparently, not very "rusty" (i can see why))
  - to do this we make `domain`, and all `validation` methods generic over `AsEntry` and `AsEncodedEntry`
- [x] remove `EntryWithOperation` trait which is no longer needed
- [x] better docs to reflect all this
- [x] more tests, especially in `DocumentStore` 

closes: #468 ` aquadoggo: [#149](https://github.com/p2panda/aquadoggo/issues/149) [#203](https://github.com/p2panda/aquadoggo/issues/203)

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header
- [x] Check if descriptions and terminology match `handbook` content (and visa-versa)

### CODE REVIEW

File: p2panda-rs/src/storage_provider/traits/document_store.rs [hunk 1]
```diff
@@ -2,55 +2,40 @@
...
-use crate::document::{Document, DocumentId, DocumentView, DocumentViewId};
+use crate::document::traits::AsDocument;
+use crate::document::{DocumentId, DocumentViewId};
 use crate::schema::SchemaId;
 use crate::storage_provider::error::DocumentStorageError;
 
-/// Storage traits for documents and document views.
+use super::OperationStore;
+
+/// Interface for querying `Documents` from the store.
```
Comment (sandreae) [inline p2panda-rs/src/storage_provider/traits/document_store.rs:12]: I'm wondering if removing `insert_docment_view` from this store trait was actually a mistake... It is very neat to only have the API outlining the query methods for documents, but now already storage provider implementations will diverge in their expected behaviour. For example, the SQL store in aquadoggo requires an extra materialisation step after insertion of operations, but `MemoryStore` doesn't. 

Realise this is a "future" problem, but makes me wonder if this design decision is right.

**Review by adzialocha — CHANGES_REQUESTED** [review]
Thank you for this massive rework! Looks and feels like a much better API now without the "optional" materialisation layer.

I mostly have comments on formulations and formatting, love the docstrings! The expressive trait signatures are a little unusual to read, but maybe that's just how it is?

With these small changes + CHANGELOG.md entry I think its good to go!

File: p2panda-rs/src/lib.rs [hunk 2]
```diff
@@ -110,6 +110,15 @@ pub trait Human {
     fn display(&self) -> String;
 }
 
+/// Trait used by p2panda structs which contain at least one id.
+///
+/// A single struct may have several id's, common use cases will be `WithId<OperationId>`,
+/// `WithId<DocumentId>` and `WithId<SchemaId>`.
```
[the remark this answers is not in the document]
Comment (sandreae) [reply p2panda-rs/src/lib.rs:116]: Yeh, they might have `WithId<OperationId>` and `WithId<SchemaId>`, does that make sense?

File: p2panda-rs/src/document/document.rs [hunk 3]
```diff
@@ -604,7 +587,7 @@ mod tests {
             Schema::get_system(SchemaId::SchemaFieldDefinition(1))
                 .unwrap()
                 .to_owned(),
-            Some(create_operation.id().to_owned().into()),
+            Some(WithId::<OperationId>::id(&create_operation).clone().into()),
```
[the remark this answers is not in the document]
Comment (sandreae) [reply p2panda-rs/src/document/document.rs:590]: Yeh, I'm not sure either... I do like it, and don't think it's weird or anything in the world of rust, but is it totally right for us, not sure.... :thinking: shall we leave it "for now"?

File: p2panda-rs/src/storage_provider/traits/operation_store.rs [hunk 4]
```diff
@@ -4,20 +4,35 @@ use async_trait::async_trait;
...
 use crate::storage_provider::error::OperationStorageError;
+use crate::WithId;
 
-/// Trait which handles all storage actions relating to `Operation`s.
+/// Storage interface for storing and querying `Operations`.
 ///
-/// This trait should be implemented on the root storage provider struct. It's definitions make up
-/// the required methods for inserting and querying operations from storage.
+/// `Operations` are a core data type of p2panda, every `Operation` is associated with one `Document``
+/// and one `PublicKey`, they form a mutation graph used to build current or historic `Document` state.
```
[the remark this answers is not in the document]
Comment (sandreae) [reply p2panda-rs/src/storage_provider/traits/operation_store.rs:16]: Haha, probably not, I thought it was more generic than "operation graph" but maybe not....

File: p2panda-rs/src/document/document.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sandreae) [reply p2panda-rs/src/document/document.rs:590]: I think most of the time the type will be inferred from the calling context.

File: p2panda-rs/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (adzialocha) [reply p2panda-rs/src/lib.rs:116]: Ah okay! Cool.

**Comment by adzialocha** [comment]
@sandreae, yeessss, one last thing maybe: Think the CHANGELOG can be more descriptive, like, what got changed? Especially as it introduces quite big API changes in this instance (in the future we would do a major release because of such change)

**Pull request merged**